subt/tools/ignstate.py

```diff
@@ -0,0 +1,170 @@
+""""
+  Extract robot positions from Ignition SQLite log file
+"""
+import cv2
+import numpy as np
+import re
+import sqlite3
+import xml.etree.ElementTree as ET
+
+from datetime import timedelta
+
+from ignition.msgs.pose_v_pb2 import Pose_V
+
+COLORS = [
+    (255, 255, 255),
+    (0, 255, 0),
+    (0, 0, 255),
+    (255, 0, 0),
+    (0, 255, 255),
+    (255, 0, 255),
+    (255, 255, 0),
+]
+SCALE = 10  # 1 pixel is 1dm
+BORDER_PX = 10  # extra border
+
+MARKERS = {
+    'backpack': (cv2.MARKER_SQUARE, 3),
+    'rescue_randy': (cv2.MARKER_DIAMOND, 2),
+    'gas': (cv2.MARKER_TRIANGLE_UP, 1),
+    'vent': (cv2.MARKER_STAR, 4),
+    'phone': (cv2.MARKER_TILTED_CROSS, 5),
+    'artifact_origin': (cv2.MARKER_CROSS, 6),
+}
+
+
+def read_poses(filename, seconds=3700):
+    ret = []
+
+    con = sqlite3.connect(filename)
+    cursor = con.cursor()
+    cursor.execute(r"SELECT id FROM topics where name LIKE '%/dynamic_pose/info';")
+    result = cursor.fetchone()
+    dynamic_topic_id = result[0]
+
+    poses = Pose_V()
+    try:
+        # cannot use WHERE filtering since the state.log is always corrupted
+        cursor.execute("SELECT message, topic_id FROM messages")
+        for m, topic_id in cursor:
+            if topic_id != dynamic_topic_id:
+                continue
+            poses.ParseFromString(m)
+            timestamp = timedelta(seconds=poses.header.stamp.sec, microseconds=poses.header.stamp.nsec / 1000)
+            if timestamp > timedelta(seconds=seconds):
+                return ret
+            current = dict()
+            for pose in poses.pose:
+                if "_" in pose.name:
+                    continue
+                current[pose.name] = pose.position
+            if len(current) > 0:
+                ret.append((timestamp, current))
+    except sqlite3.DatabaseError as e:
+        print(f"{type(e).__name__}: {e}")
+
+    return ret
+
+
+def read_artifacts(filename):
+    ret = []
+    con = sqlite3.connect(filename)
+    cursor = con.cursor()
+    cursor.execute(f"SELECT id FROM topics WHERE name == '/logs/sdf'")
+    result = cursor.fetchone()
+    sdf_id = result[0]
+    cursor.execute(r"SELECT message, topic_id FROM messages")
+    for message, topic_id in cursor:
+        if topic_id == sdf_id:
+            world = message
+            break
+    root = ET.fromstring(world[4:])
+    type_re = re.compile('^(backpack|rescue_randy|gas|vent|phone|artifact_origin)')
+    for model in root.iterfind("./world/model"):
+        name = model.get('name')
+        match = type_re.match(name)
+        if match:
+            kind = match.group(1)
+            x, y, z = [float(a) for a in model.find('./pose').text.split()[:3]]
+            ret.append([kind, [x,y,z]])
+    return ret
+
+
+def draw(poses, artifacts):
+    min_x, min_y = 10_000, 10_000
```

Comment:
why is here underscore? (10_000)

Reply:
I always have trouble counting the zeros. This makes it (at least for me) visually easier to confirm what the value is.

Reply:
OK, I did not know that you can add `_` in numbers, thanks

Reply:
[PEP 515 -- Underscores in Numeric Literals](https://www.python.org/dev/peps/pep-0515/)